scalapact-core/src/main/scala/com/itv/scalapactcore/common/matching/InteractionMatchingPrograms.scala

```diff
@@ -0,0 +1,213 @@
+
+  def matchResponseProgram(expected: InteractionResponse, actual: InteractionResponse): MatchResponseStage[ResponseMatchOutcome] =
+    for {
+      s <- matchResponseStatus(expected, actual)
+      h <- matchResponseHeaders(expected, actual)
+      b <- matchResponseBody(expected, actual)
+    } yield ResponseMatchOutcome(s, h, b)
+
+  type MatchRequestStage[A] = Free[MatchRequestStageA, A]
+
+  // Request
+  private def matchRequestMethod(expected: InteractionRequest, actual: InteractionRequest): MatchRequestStage[MatchMethodOutcome] =
+    Free.liftF(MatchRequestMethod(expected, actual))
+
+  private def matchRequestPath(expected: InteractionRequest, actual: InteractionRequest): MatchRequestStage[MatchPathOutcome] =
+    Free.liftF(MatchRequestPath(expected, actual))
+
+  private def matchRequestHeaders(expected: InteractionRequest, actual: InteractionRequest): MatchRequestStage[MatchHeadersOutcome] =
+    Free.liftF(MatchRequestHeaders(expected, actual))
+
+  private def matchRequestBody(expected: InteractionRequest, actual: InteractionRequest): MatchRequestStage[MatchBodyOutcome] =
+    Free.liftF(MatchRequestBody(expected, actual))
+
+  def matchRequestProgram(expected: InteractionRequest, actual: InteractionRequest): MatchRequestStage[RequestMatchOutcome] =
+    for {
+      m <- matchRequestMethod(expected, actual)
+      p <- matchRequestPath(expected, actual)
+      h <- matchRequestHeaders(expected, actual)
+      b <- matchRequestBody(expected, actual)
+    } yield RequestMatchOutcome(m, p, h, b)
+
+}
+
+object MatchingInterpreters {
+
+  object Request {
+
+    object permissive extends (MatchRequestStageA ~> Id) {
+
+      def apply[A](stage: MatchRequestStageA[A]): Id[A] = {
+        stage match {
+          case MatchRequestMethod(expected, actual) =>
+            if(MethodMatching.matchMethods(expected.method)(actual.method)) MatchMethodSuccess
+            else MatchMethodFailed
+
+          case MatchRequestPath(expected, actual) =>
+            if(PathMatching.matchPaths(PathAndQuery(expected.path, expected.query))(PathAndQuery(actual.path, actual.query))) MatchPathSuccess
+            else MatchPathFailed
+
+          case MatchRequestHeaders(expected, actual) =>
+            if(HeaderMatching.matchHeaders(expected.matchingRules)(expected.headers)(actual.headers)) MatchHeadersSuccess
+            else MatchHeadersFailed
+
+          case MatchRequestBody(expected, actual) =>
+            if(BodyMatching.matchBodies(expected.matchingRules)(expected.body)(actual.body)) MatchBodySuccess
+            else MatchBodyFailed
+
+        }
+      }
+
+    }
+
+    object strict extends (MatchRequestStageA ~> Id) {
+
+      def apply[A](stage: MatchRequestStageA[A]): Id[A] = {
+        stage match {
+          case MatchRequestMethod(expected, actual) =>
+            if(MethodMatching.matchMethods(expected.method)(actual.method)) MatchMethodSuccess
+            else MatchMethodFailed
```

Comment:
To elaborate. Intellij doesn't understand that `Unit` or `List[String]` which are clearly two different possible return types, could both be equal to `A` in the same function. You can make Intellij happy by appending `.asInstanceOf[A]` but that makes me unhappy.